### PR TITLE
CFP AVAD: Adds new FeedRange to ChangeFeedProcessorContext

### LINKED ISSUE #4620: CFP AVAD: Add public access to FeedRange from ChangeFeedProcessorContext
**Is your feature request related to a problem? Please describe.**
We have a customer that needs access to this for a complex scenario where they need to persist feed ranges as "bookmarks" along with the changed document's LSN. This is later validated to confirm whether the LSN has been processed or not. Details of the holistic view is located [here](https://github.com/Azure/azure-cosmos-dotnet-v3/issues/4483).

Bookmark sample:

```json
[
	{
		"Range": {
			"min": "",
			"max": "05C1DFFFFFFFFC"
		},
		"LSN": "0"
	}
]
```
**Describe the solution you'd like**
Give the customer access to _FeedRange_ from _ChangeFeedProcessorContext_

`ChangeFeedProcessorContext.FeedRange`

**Describe alternatives you've considered**
n/A

**Additional context**
N/A

### PR DESCRIPTION
**Pull request by philipthomas-MSFT**
# Pull Request Template
Details on #4620 

## Description

Please include a summary of the change and which issue is fixed. Include samples if adding new API, and include relevant motivation and context. List any dependencies that are required for this change.

## Type of change

Please delete options that are not relevant.

- [] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [] This change requires a documentation update

## Closing issues

To automatically close an issue: closes #4620

### CODE REVIEW

**Comment by philipthomas-MSFT** [comment]
Blocking myself. Some other tests are failing and I need to run UpdateContracts.ps1

**Comment by philipthomas-MSFT** [comment]
> Blocking myself. Some other tests are failing and I need to run UpdateContracts.ps1

Fixed

File: Microsoft.Azure.Cosmos/src/ChangeFeedProcessor/FeedProcessing/ProcessorOptions.cs [hunk 1]
```diff
@@ -22,5 +21,7 @@ internal class ProcessorOptions
         public DateTime? StartTime { get; set; }
 
         public TimeSpan RequestTimeout { get; set; } = CosmosHttpClient.GatewayRequestTimeout;
+
+        public FeedRangeInternal FeedRange { get; set; }
```
Comment (ealsur) [inline Microsoft.Azure.Cosmos/src/ChangeFeedProcessor/FeedProcessing/ProcessorOptions.cs:25]: No need for it to be FeedRangeInternal?

```suggestion
        public FeedRange FeedRange { get; set; }
```

File: Microsoft.Azure.Cosmos/tests/Microsoft.Azure.Cosmos.EmulatorTests/ChangeFeed/GetChangeFeedProcessorBuilderWithAllVersionsAndDeletesTests.cs [hunk 2]
```diff
@@ -573,20 +576,86 @@ await GetChangeFeedProcessorBuilderWithAllVersionsAndDeletesTests
...
                 throughput: 10000,
                 cancellationToken: this.cancellationToken);
 
-            return (ContainerInternal)response;
+            return (ContainerInternal)this.containerResponse;
+        }
+
+        [TestMethod]
+        [Owner("philipthomas-MSFT")]
+        [Description("Scenario: Test to confirm that FeedRange is coming back in the context.")]
```
Comment (ealsur) [inline Microsoft.Azure.Cosmos/tests/Microsoft.Azure.Cosmos.EmulatorTests/ChangeFeed/GetChangeFeedProcessorBuilderWithAllVersionsAndDeletesTests.cs:597]: nit: This is not just for AVAD, right? Can we simply add this check to any of the existing tests? That way we reduce the tests run time.

For example: 

https://github.com/Azure/azure-cosmos-dotnet-v3/blob/29a26125baefb5723b5a7bba82e97f1b96c616ee/Microsoft.Azure.Cosmos/tests/Microsoft.Azure.Cosmos.EmulatorTests/ChangeFeed/DynamicTests.cs#L383-L391

File: Microsoft.Azure.Cosmos/tests/Microsoft.Azure.Cosmos.EmulatorTests/ChangeFeed/GetChangeFeedProcessorBuilderWithAllVersionsAndDeletesTests.cs [hunk 3]
```diff
@@ -573,20 +576,86 @@ await GetChangeFeedProcessorBuilderWithAllVersionsAndDeletesTests
...
+            {
+                Assert.Fail(exception.ToString());
+            }
+        }
+
+        private static async Task ValidateFeedRangeAsync(CosmosClient cosmosClient, FeedRange feedRange, string containerRid)
+        {
+            Assert.IsNotNull(feedRange);
+
+            Routing.PartitionKeyRangeCache partitionKeyRangeCache = await cosmosClient.DocumentClient.GetPartitionKeyRangeCacheAsync(NoOpTrace.Singleton);
```
Comment (ealsur) [inline Microsoft.Azure.Cosmos/tests/Microsoft.Azure.Cosmos.EmulatorTests/ChangeFeed/GetChangeFeedProcessorBuilderWithAllVersionsAndDeletesTests.cs:649]: You can just use `Container.GetPartitionKeyRangesAsync` with the FeedRange and using the Monitored Container.

**Review by ealsur — COMMENTED** [review]
Looks good, just some small polishing comments

File: Microsoft.Azure.Cosmos/tests/Microsoft.Azure.Cosmos.EmulatorTests/ChangeFeed/GetChangeFeedProcessorBuilderWithAllVersionsAndDeletesTests.cs (hunk 2 above)
Comment (philipthomas-MSFT) [reply Microsoft.Azure.Cosmos/tests/Microsoft.Azure.Cosmos.EmulatorTests/ChangeFeed/GetChangeFeedProcessorBuilderWithAllVersionsAndDeletesTests.cs:597]: I'll consider. Thanks.

File: Microsoft.Azure.Cosmos/tests/Microsoft.Azure.Cosmos.EmulatorTests/ChangeFeed/GetChangeFeedProcessorBuilderWithAllVersionsAndDeletesTests.cs (hunk 3 above)
Comment (philipthomas-MSFT) [reply Microsoft.Azure.Cosmos/tests/Microsoft.Azure.Cosmos.EmulatorTests/ChangeFeed/GetChangeFeedProcessorBuilderWithAllVersionsAndDeletesTests.cs:649]: Cool. That will make this easier.

File: Microsoft.Azure.Cosmos/src/ChangeFeedProcessor/ChangeFeedProcessorContext.cs [hunk 4]
```diff
@@ -23,5 +23,10 @@ public abstract class ChangeFeedProcessorContext
         /// Gets the headers related to the service response that provided the changes.
         /// </summary>
         public abstract Headers Headers { get; }
+
+        /// <summary>
+        /// Gets the feed range.
+        /// </summary>
+        public abstract FeedRange FeedRange { get; }
```
Comment (kirankumarkolli) [inline Microsoft.Azure.Cosmos/src/ChangeFeedProcessor/ChangeFeedProcessorContext.cs:30]: Please update description with issues context (will be quite helpful for CX when navigating the changelog)

Comment (philipthomas-MSFT) [reply Microsoft.Azure.Cosmos/src/ChangeFeedProcessor/ChangeFeedProcessorContext.cs:30]: @kirankumarkolli  I am updating the other draft PR with the whole view of the problem. I was asked to break the PRs up into multiple PRs so this one has the details for just FeedRange on the ChangeFeedProcessorContext. The other one that has FindOverlappingRanges will have all the scenarios and diagrams once I am done.

Comment (philipthomas-MSFT) [reply Microsoft.Azure.Cosmos/src/ChangeFeedProcessor/ChangeFeedProcessorContext.cs:30]: So if you want to read this one, it is in the description, but I will link it here also.

https://github.com/Azure/azure-cosmos-dotnet-v3/issues/4620

Comment (kirankumarkolli) [reply Microsoft.Azure.Cosmos/src/ChangeFeedProcessor/ChangeFeedProcessorContext.cs:30]: Issue description is great, I did went over it before. 

As pattern the PR description needs to carry full context (before and after, design choices etc...)
This will help long term.

File: Microsoft.Azure.Cosmos/src/ChangeFeedProcessor/ChangeFeedProcessorContext.cs [hunk 5]
```diff
@@ -23,5 +23,10 @@ public abstract class ChangeFeedProcessorContext
         /// Gets the headers related to the service response that provided the changes.
         /// </summary>
         public abstract Headers Headers { get; }
+
+        /// <summary>
+        /// Gets the feed range.
```
Comment (bartelink) [inline Microsoft.Azure.Cosmos/src/ChangeFeedProcessor/ChangeFeedProcessorContext.cs:28]: No undocumentation please!

Comment (ealsur) [reply Microsoft.Azure.Cosmos/src/ChangeFeedProcessor/ChangeFeedProcessorContext.cs:28]: What about:

```suggestion
        /// Gets the monitored container's <see cref="FeedRange"/> from which the changes come from.
```

Comment (bartelink) [reply Microsoft.Azure.Cosmos/src/ChangeFeedProcessor/ChangeFeedProcessorContext.cs:28]:
```suggestion
The <see cref="FeedRange"/> within the monitored container from which the changes emanated.
```
(mainly rewording to remove two 'from's in Matias's suggestion). (Or remove the first 'within' and end on 'arose within' ?)

File: Microsoft.Azure.Cosmos/tests/Microsoft.Azure.Cosmos.EmulatorTests/ChangeFeed/DynamicTests.cs [hunk 6]
```diff
@@ -380,7 +382,7 @@ public async Task TestWithStartTime_CustomTime()
             Assert.AreEqual("doc5.doc6.doc7.doc8.doc9.", accumulator);
         }
 
-        private void ValidateContext(ChangeFeedProcessorContext changeFeedProcessorContext)
+        private async void ValidateContext(ChangeFeedProcessorContext changeFeedProcessorContext)
```
Comment (kirankumarkolli) [inline Microsoft.Azure.Cosmos/tests/Microsoft.Azure.Cosmos.EmulatorTests/ChangeFeed/DynamicTests.cs:385]: Avoid `async void` https://learn.microsoft.com/en-us/archive/msdn-magazine/2013/march/async-await-best-practices-in-asynchronous-programming#avoid-async-void

Comment (philipthomas-MSFT) [reply Microsoft.Azure.Cosmos/tests/Microsoft.Azure.Cosmos.EmulatorTests/ChangeFeed/DynamicTests.cs:385]: Ya, @kirankumarkolli, I was on the fence about this one. Because @ealsur requested me to add this validation here instead of having my own test. I will have to change more on these tests making everything async.

File: Microsoft.Azure.Cosmos/tests/Microsoft.Azure.Cosmos.EmulatorTests/ChangeFeed/GetChangeFeedProcessorBuilderWithAllVersionsAndDeletesTests.cs [hunk 7]
```diff
@@ -588,53 +588,5 @@ private async Task<ContainerInternal> CreateMonitoredContainer(ChangeFeedMode ch
 
             return (ContainerInternal)response;
         }
-
```
Comment (kirankumarkolli) [inline Microsoft.Azure.Cosmos/tests/Microsoft.Azure.Cosmos.EmulatorTests/ChangeFeed/GetChangeFeedProcessorBuilderWithAllVersionsAndDeletesTests.cs:591]: how is this change related? Also why are these tests deleted (just added in previous PR)

File: Microsoft.Azure.Cosmos/tests/Microsoft.Azure.Cosmos.EmulatorTests/ChangeFeed/DynamicTests.cs (hunk 6 above)
Comment (philipthomas-MSFT) [reply Microsoft.Azure.Cosmos/tests/Microsoft.Azure.Cosmos.EmulatorTests/ChangeFeed/DynamicTests.cs]: fixed @kirankumarkolli

File: Microsoft.Azure.Cosmos/tests/Microsoft.Azure.Cosmos.EmulatorTests/ChangeFeed/GetChangeFeedProcessorBuilderWithAllVersionsAndDeletesTests.cs (hunk 7 above)
Comment (philipthomas-MSFT) [reply Microsoft.Azure.Cosmos/tests/Microsoft.Azure.Cosmos.EmulatorTests/ChangeFeed/GetChangeFeedProcessorBuilderWithAllVersionsAndDeletesTests.cs:591]: Because @kirankumarkolli , I split the PR into multiple PRs and this was already added on the other one that is complete.

Comment (philipthomas-MSFT) [reply Microsoft.Azure.Cosmos/tests/Microsoft.Azure.Cosmos.EmulatorTests/ChangeFeed/GetChangeFeedProcessorBuilderWithAllVersionsAndDeletesTests.cs:591]: > Because @kirankumarkolli , I split the PR into multiple PRs and this was already added on the other one that is complete.

ok. wait. maybe a bad merge. because i was asked to separate this, the other PR merged it into master, but this is the first PR where I moved it.

File: Microsoft.Azure.Cosmos/tests/Microsoft.Azure.Cosmos.EmulatorTests/ChangeFeed/DynamicTests.cs [hunk 8]
```diff
@@ -61,7 +61,7 @@ public async Task TestWithRunningProcessor()
                         allDocsProcessed.Set();
                     }
 
-                    return Task.CompletedTask;
```
Comment (kirankumarkolli) [inline Microsoft.Azure.Cosmos/tests/Microsoft.Azure.Cosmos.EmulatorTests/ChangeFeed/DynamicTests.cs:64]: return not needed

File: Microsoft.Azure.Cosmos/tests/Microsoft.Azure.Cosmos.EmulatorTests/ChangeFeed/DynamicTests.cs [hunk 9]
```diff
@@ -389,6 +389,19 @@ private void ValidateContext(ChangeFeedProcessorContext changeFeedProcessorConte
...
+
+        private async Task ValidateFeedRangeAsync(FeedRange feedRange)
+        {
+            Assert.IsNotNull(feedRange);
+            
+            IEnumerable<string> partitionKeyRanges = await this.Container.GetPartitionKeyRangesAsync(feedRange);
+
+            Assert.IsNotNull(partitionKeyRanges);
+            Assert.AreEqual(1, partitionKeyRanges.Count());
+            Assert.AreEqual(expected: "0", actual: partitionKeyRanges.FirstOrDefault());
```
Comment (ealsur) [inline Microsoft.Azure.Cosmos/tests/Microsoft.Azure.Cosmos.EmulatorTests/ChangeFeed/DynamicTests.cs:404]: nit: Not sure if it always be partition key range id 0, if the test is creating multiple partitions, this check might fail.

**Review by ealsur — APPROVED** [review]

**Review by kirankumarkolli — APPROVED** [review]

**Review by kirankumarkolli — APPROVED** [review]

**Review by ealsur — APPROVED** [review]
